dht/dht.go

```diff
@@ -618,13 +648,28 @@ func (s *Server) nodeByID(id string) *node {
 	return nil
 }
 
-func (s *Server) handleQuery(source dHTAddr, m Msg) {
+func (s *Server) handleQuery(source DHTAddr, m Msg) {
+	var (
+		newmsg *Msg
+		skip   bool
+	)
 	node := s.getNode(source, m.SenderID())
 	node.lastGotQuery = time.Now()
 	// Don't respond.
 	if s.config.Passive {
 		return
 	}
+	// Call any hooks that may apply
+	if hook, ok := s.hooks[m.Q]; ok {
+		newmsg, skip = hook(&source, node, &m)
+		if newmsg != nil {
```

Comment:
I think you're better off allowing the callee to modify the Msg in place, and telling them that any changes they make will be propagated.